syntax = "proto2";

package efos;

message TypeDogma {
    message TypeDogmaEntry {
        message DogmaAttributes {
            required int32 attributeID = 1;
            required double value = 2;
        }

        message DogmaEffects {
            required int32 effectID = 1;
            required bool isDefault = 2;
        }

        repeated DogmaAttributes dogmaAttributes = 1;
        repeated DogmaEffects dogmaEffects = 2;
    }

    map<int32, TypeDogmaEntry> entries = 1;
}

message Types {
    message Type {
        required int32 groupID = 1;
        required int32 categoryID = 2;
        required bool published = 3;

        optional double capacity = 4;
        optional double mass = 5;
        optional double radius = 6;
        optional double volume = 7;
    }

    map<int32, Type> entries = 1;
}

message Categories {
    message Category {
        required string name = 1;
        required bool published = 2;
    }

    map<int32, Category> entries = 1;
}

message Groups {
    message Group {
        required string name = 1;
        required int32 categoryID = 2;
        required bool published = 3;
    }

    map<int32, Group> entries = 1;
}

message MarketGroups {
    message MarketGroup {
        required string name = 1;

        optional int32 parentGroupID = 2;
        optional int32 iconID = 3;
    }

    map<int32, MarketGroup> entries = 1;
}

message DogmaAttributes {
    message DogmaAttribute {
        required string name = 1;
        required bool published = 2;
        required double defaultValue = 3;
        required bool highIsGood = 4;
        required bool stackable = 5;
    }

    map<int32, DogmaAttribute> entries = 1;
}

message DogmaEffects {
    message DogmaEffect {
        message ModifierInfo {
            enum Domain {
                itemID = 0;
                shipID = 1;
                charID = 2;
                otherID = 3;
                structureID = 4;
                target = 5;
                targetID = 6;
            }

            enum Func {
                ItemModifier = 0;
                LocationGroupModifier = 1;
                LocationModifier = 2;
                LocationRequiredSkillModifier = 3;
                OwnerRequiredSkillModifier = 4;
                EffectStopper = 5;
            }

            required Domain domain = 1;
            required Func func = 2;
            optional int32 modifiedAttributeID = 3;
            optional int32 modifyingAttributeID = 4;
            optional int32 operation = 5;
            optional int32 groupID = 6;
            optional int32 skillTypeID = 7;
        }

        required string name = 1;
        required int32 effectCategory = 2;
        repeated ModifierInfo modifierInfo = 3;
    }

    map<int32, DogmaEffect> entries = 1;
}

message BuffCollections {
    message Buff {
        enum AggregateMode {
            MAXIMUM = 0;
            MINIMUM = 1;
        }

        enum ShowOutputValueInUI {
            SHOW_NORMAL = 0;
            SHOW_INVERTED = 1;
            HIDE = 2;
        }

        enum OperationName {
            POST_MUL = 0;
            POST_PERCENT = 1;
            MOD_ADD = 2;
            POST_ASSIGNMENT = 3;
        }

        message ItemModifier {
            required int32 dogmaAttributeID = 1;
        }

        message LocationGroupModifier {
            required int32 dogmaAttributeID = 1;
            required int32 groupID = 2;
        }

        message LocationRequiredSkillModifier {
            required int32 dogmaAttributeID = 1;
            required int32 skillID = 2;
        }

        required AggregateMode aggregateMode = 1;
        required int32 buffID = 2;
        required string developerDescription = 3;
        repeated ItemModifier itemModifiers = 4;
        repeated LocationGroupModifier locationGroupModifiers = 5;
        repeated ItemModifier locationModifiers = 6;
        repeated LocationRequiredSkillModifier locationRequiredSkillModifiers = 7;
        required OperationName operationName = 8;
        required ShowOutputValueInUI showOutputValueInUI = 9;

        optional string displayName = 10;
    }

    map<int32, Buff> entries = 1;
}
